tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs

```diff
@@ -40,145 +64,104 @@ public MatchedCodeOwnerEntry(List<CodeOwnerEntry> entries, string targetPath)
-            {
-                codeownersPath += "**";
-            }
-            else
+            // CODEOWNERS paths that do not start with "/" are relative and considered invalid.
+            // However, here we handle such cases to accomodate for parsing CODEOWNERS file
+            // paths that somehow slipped through validation. We do so by instead treating
+            // such paths as if they were absolute to repository root, i.e. starting with "/".
+            if (!codeownersPath.StartsWith("/"))
+                codeownersPath = "/" + codeownersPath;
+
+            string pattern = codeownersPath;
+
+            // Kind of hoping here that the CODEOWNERS path will never have
+            // "_DOUBLE_STAR_" or "_SINGLE_STAR_" strings in it.
+            pattern = pattern.Replace("**", "_DOUBLE_STAR_");
+            pattern = pattern.Replace("*", "_SINGLE_STAR_");
+
+            pattern = Regex.Escape(pattern);
+
+            // Denote that all paths are absolute by prepending "beginning of string" symbol.
+            pattern = "^" + pattern;
+
+            // Lack of slash at the end denotes the path is a path to a file,
+            // per our validation logic.
+            // Note we assume this is the case even if the path is invalid,
+            // even though in such case it might not necessarily be true.
+            if (!(pattern.EndsWith("/") 
+                  || pattern.EndsWith("_DOUBLE_STAR_")))
             {
-                // codeownersPath directory already has stars in the suffix, so nothing to do.
-                // Example paths:
-                // apps/*/
-                // apps/**/
+                // Append "end of string", symbol, denoting the match has to be exact,
+                // not a substring, as we are dealing with a file.
+                pattern += "$";
             }
 
-            return codeownersPath;
-        }
-
-        private static PatternMatchingResult MatchGlobPatterns(
-            string targetPath,
-            List<string> patterns)
-        {
-            // Note we use StringComparison.Ordinal, not StringComparison.OrdinalIgnoreCase,
-            // as CODEOWNERS paths are case-sensitive.
-            var globMatcher = new Matcher(StringComparison.Ordinal);
-
-            foreach (var pattern in patterns)
-            {
-                globMatcher.AddInclude(pattern);
-            }
-            
-            var dir = new InMemoryDirectoryInfo(
-                // This 'rootDir: "/"' is used here only because the globMatcher API requires it.
-                rootDir: "/", 
-                files: new List<string> { targetPath });
-
-            var patternMatchingResult = globMatcher.Execute(dir);
-            return patternMatchingResult;
+            pattern = pattern.Replace("_DOUBLE_STAR_/", "(.*)");
+            pattern = pattern.Replace("/_DOUBLE_STAR_", "(.*)");
+            pattern = pattern.Replace("_DOUBLE_STAR_", "(.*)");
```

Comment:
What about `/_DOUBLE_STAR_/`? Do we need to preserve any `/` in this double star case?

Reply:
We will preserve one `/` because `/_DOUBLE_STAR_/` case will be replaced by `_DOUBLE_STAR_/` case, preserving the prefix `/`. I [added comments explaining that](https://github.com/Azure/azure-sdk-tools/blob/5aaa156e0eaaecdda673ae7663262cf7a921ad2f/tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs#L166-L172). 

Your remark also made me realize [I was missing a test case](https://github.com/Azure/azure-sdk-tools/blob/5aaa156e0eaaecdda673ae7663262cf7a921ad2f/tools/code-owners-parser/Azure.Sdk.Tools.CodeOwnersParser.Tests/CodeOwnersFileTests.cs#L99) for the `_DOUBLE_START` case. Thanks!